File: leetcode/medium/0003-longest-substring-without-repeating-characters/solution.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        last_seen = [-1] * 128

        left = 0
        max_len = 0

        for right in range(len(s)):

            ch = ord(s[right])

            if last_seen[ch] >= left:
                left = last_seen[ch] + 1

            last_seen[ch] = right

            max_len = max(max_len, right - left + 1)

        return max_len

File: leetcode/medium/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int[] lastSeen = new int[128];

        // Initialize all positions to -1
        Arrays.fill(lastSeen, -1);

        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {

            char ch = s.charAt(right);

            // If character is inside current window,
            // move left pointer just after its previous occurrence.
            if (lastSeen[ch] >= left) {
                left = lastSeen[ch] + 1;
            }

            lastSeen[ch] = right;

            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}